feat(scheduler): check if elevator should stop

braedenkloke opened the pull request:
# Summary
Implemented check in method isElevatorStopping.

# Implementation Details
* Add field floorRequests to scheduler as instructed by @orselane 
* implement method getElevatorDirection() to get elevator direction from a ElevatorStateEvent (high potential for reusability!)

# Testing
* Wrote positive unit test for isElevatorStopping.
* created test class SchedulerTest nested in Scheduler to enable testing

braedenkloke committed:
docs(scheduler): add java doc for isElevatorStopping

braedenkloke committed:
test(scheduler): initialize unit tests for scheduler

braedenkloke committed:
test(scheduler): write setup for isElevatorStoppingTest

Refactor isElevatorStopping to public method.

braedenkloke committed:
feat(scheduler): add field to hold a collection of all floor requests

Implemented as List.

braedenkloke committed:
docs(scheduler): add design decisions to processDestinationEvent

Additionally, refactor JavaDocs to simplify WHAT processDestinationEvent does.

braedenkloke committed:
test(scheduler): initialize positive test for is elevator stopping.

Test initalized to fail as per TDD (Red-Green-Refactor).

braedenkloke committed:
feat(scheduler): partially implement isElevatorStopping

Implement check for if elevator.currentFloor is in list of destination floors.

Further work involves implementing private functions to take the union of two dest event
list and to reduce a list of dest events to a list of dest floors.
See TODO items in method isElevatorStopping.

braedenkloke committed:
feat(scheduler): implement filterDestinationFloors

Method filterDestinationFloors filters a list of destination events
to a list of destination floors. This method is used in method isElevatorStopping.

braedenkloke committed:
feat(scheduler): implement isElevatorStopping

Implemented private method unionOfDestinationEvents. Last piece of the puzzle for
implementing isElevatorStopping.

braedenkloke committed:
refactor(scheduler.isElevatorStopping): changed arg event to e

braedenkloke committed:
refactor(schedulerTest): move docs from header to method body

braedenkloke committed:
refactor(scheduler): change method FROM filterDestinationFloors TO toDestinationFloors

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,57 @@
     private final DMA_Transmitter transmitterToFloor;
     private final DMA_Transmitter transmitterToElevator;
     private final DMA_Receiver receiver;
+    private List<DestinationEvent> destinationEvents;
```
This should be a set, in any case it shouldn't hold any duplicates because multiple request to the same floor should be handled the same. It also helps for the set math we want to do. 
I'd suggest naming it "floor requests" for a more descriptive name too.

braedenkloke replied:
Resolved: 49c63f49dced61319f27bc3ee5155f5317a0ce75

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,57 @@
      */
-    private void processDestinationEvent(DestinationEvent destinationEvent) {
-        transmitterToElevator.send(destinationEvent);
+    public void processDestinationEvent(DestinationEvent destinationEvent) {
```
This should not be public, we should not be able to access the internal scheduler behavior from outside the class otherwise we could bypass our state machine.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,57 @@
+        // TODO(@alex): describe WHY event is being forwarded to elevator subsystem
+        //   HEY ALEX! It would be helpful for the team for you to document HERE why the
+        //   elevator subsystem is getting a duplicated event. Much love, Braeden.
+        transmitterToElevator.send(destinationEvent);
```
```suggestion
```
Remnant of iteration 1, it should not be getting the floor requests.

braedenkloke replied:
Resolved: 0e936882b8d28e73dcfdd3605812e92d6e0ece83

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,57 @@
+        //
+        // See state diagram of scheduler for additional context.
+
+        List<DestinationEvent> union = unionOfDestinationEvents(destinationEvents, e.destinationEvents());
```
I'll rework "destinationEvents" field from the event to be a map later, I forgot about it sorry.

braedenkloke replied:
Resolved. Merge code from PR#38 and applied fixes.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,57 @@
+
+        List<DestinationEvent> union = unionOfDestinationEvents(destinationEvents, e.destinationEvents());
+        List<Integer> destinationFloors = toDestinationFloors(union);
+        if (destinationFloors.contains(e.currentFloor())) {
```
```suggestion
        return destinationFloors.contains(e.currentFloor())
```

braedenkloke replied:
Resolved: cccca1f92712abc3fc61a025ab5e661acb0eb425

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,57 @@
+        // See state diagram of scheduler for additional context.
+
+        List<DestinationEvent> union = unionOfDestinationEvents(destinationEvents, e.destinationEvents());
+        List<Integer> destinationFloors = toDestinationFloors(union);
```
Set I believe are more efficient, and more practical for our usecases.

braedenkloke replied:
Implemented as sets. Resolved.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -62,6 +89,39 @@ else if (event instanceof DestinationEvent)
+     * @param events List of destination events.
+     * @return List of destination floors.
+     */
+    private List<Integer> toDestinationFloors(List<DestinationEvent> events) {
```
This removes direction from the matching. I think if you matched with the whole event, you can include direction checking without having to filter destination events. Currently you do not check for direction at all in "isElevatorStopping" from my understanding.

orselane replied:
Direction should be derived from the passengers in the elevator, it might get removed from the event soon in the next rework.

braedenkloke replied:
Direction checking wasn't mentioned in the state diagram. Can you update the state diagram so I can implement your suggestions properly?

Scheduler State Diagram: https://app.diagrams.net/#G14DeIsZoibYy-BnQ7c81sKBTpHJQuF3Pq#%7B%22pageId%22%3A%22T1rtvHDK5Ro4PfxUw1XM%22%7D

orselane replied:
We talked about this during the meeting, Arthur brought it up. I made an amendment it's in the picture you took. Implementation wise, there's not much work to do except find the elevator direction which I mentioned in the comments above. Arthur implemented a method to do that in Elevator which you can copy over until we think of a better way to share the methods.

orselane commented on test/unit/SchedulerSubsystem/SchedulerTest.java:
```diff
@@ -0,0 +1,58 @@
+     */
+    @Test
+    @DisplayName("isElevatorStopping positive test")
+    public void testPositiveIsElevatorStopping() {
```
It's great to have a test for this method, but please take a look at this discussion board post:

https://brightspace.carleton.ca/d2l/le/224517/discussions/threads/530966/View

We should not be making private methods public purely to test them. If you would like, you can try to implement one of the methods mentioned in the discussion board to test that function.

braedenkloke replied:
isElevatorStopping is OK as public. Considering the overall team's capacity, I don't agree that the added complexity in the given suggestions are worth the effort implementing. We can leave a comment  in the JavaDocs mentioning how if we had more time we could make it private!

I've refactored the other methods to stay private (eg. processMessage). I simply added a setter method for floorRequests to enable setting state in tests, as opposed to trying to set state by publishing events!

orselane reviewed (changes requested):
I will go implement those changes to ElevatorStateEvent I mentioned in the review.

braedenkloke committed:
refactor(scheduler): scheduler does not forward dest events to elevator

Confirmed by Alex. Removed line is deprecated.

braedenkloke committed:
refactor(scheduler): eliminate if-else block in isElevatorStopping

As per Alex's suggestion.

braedenkloke committed:
fix(scheduler): make processDestinationEvent private again

Implemented setter method in Scheduler to control state for tests.

braedenkloke committed:
refactor(scheduler): change collection of destination events from list to set

Updated tests.

braedenkloke committed:
docs(scheduler): polish various java docs

And add funny comments.

braedenkloke committed:
Merge branch 'develop/latest' into develop/scheduler-check-is-elevator-stopping

braedenkloke committed:
fix(scheduler): accept new elevatorEventState format

braedenkloke committed:
fix(scheduler): pass arg to constructor to enable compiling

event.direction() is deprecated.

braedenkloke committed:
test(scheduler): fix test to accept map of elevator destination events

braedenkloke committed:
refactor(scheduler): change field FROM destinationEvents TO floorRequests

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,46 @@
+        //
+        // See state diagram of scheduler for additional context.
+
+        Set<DestinationEvent> union = unionOfDestinationEvents(floorRequests, e.passengerCountMap().keySet());
```
Use the sets method: addAll for an built-in union method.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -62,6 +79,49 @@ else if (event instanceof DestinationEvent)
+     * @param B Set of destination events.
+     * @return Union of both sets.
+     */
+    private Set<DestinationEvent> unionOfDestinationEvents(
```
Can get rid of this method, use addAll instead.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,46 @@
+
+        Set<DestinationEvent> union = unionOfDestinationEvents(floorRequests, e.passengerCountMap().keySet());
+        Set<Integer> destinationFloors = filterDestinationFloors(union);
+        return destinationFloors.contains(e.currentFloor());
```
```suggestion
        return union.contains(new DestinationEvent(e.currentFloor(), getElevatorDirection(e.passengerCountMap())));
```

orselane replied:
This will check for direction as well as floor at the same time making sure you don't serve the wrong direction.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,46 @@
+        // See state diagram of scheduler for additional context.
+
+        Set<DestinationEvent> union = unionOfDestinationEvents(floorRequests, e.passengerCountMap().keySet());
+        Set<Integer> destinationFloors = filterDestinationFloors(union);
```
```suggestion
```

orselane replied:
This get rid of the destination information which we do need to check against.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -62,6 +79,49 @@ else if (event instanceof DestinationEvent)
+     * @param events Set of destination events.
+     * @return Set of destination floors.
+     */
+    private Set<Integer> filterDestinationFloors(Set<DestinationEvent> events) {
```
This get rid of the destination information which we do need to check against, detrimental at the moment I would remove this metod.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -62,6 +79,49 @@ else if (event instanceof DestinationEvent)
+     *
+     * @param events Set of destination events to be set.
+     */
+    public void setFloorRequests(Set<DestinationEvent> events) {
```
This seems purely like a testing function, as I've said before I think that's a bad idea. Especially setters.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -8,30 +8,46 @@
+     * @param e Elevator state to check.
+     * @return True if elevator should stop, false otherwise.
+     */
+    public boolean isElevatorStopping(ElevatorStateEvent e) {
```
I'm still not a fan of keeping this public but we can talk more on Tuesday when we discuss testing.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -43,7 +59,8 @@ private void processElevatorEvent(ElevatorStateEvent event) {
         else // Keep moving
-            transmitterToElevator.send(new MoveElevatorCommand(event.elevatorNum()));
+            // Note(@braeden): event.direction() is deprecated
+            transmitterToElevator.send(new MoveElevatorCommand(event.elevatorNum(), event.direction()));
```
You should use the new direction derivation method instead of event.direction()

orselane commented:
To be clear, all the public/testing stuff is not blocking the merge in my opinion, we can discuss/fix later. The rest is.

braedenkloke committed:
docs(scheduler): add todo items

braedenkloke committed:
refactor(test): move SchedulerTest to nested test class in Scheduler

braedenkloke committed:
refactor(SchedulerTest): change FROM s TO scheduler

Improves readability.

braedenkloke committed:
refactor(scheduler): implement union of destination events using Set.addAll()

Deleted deprecated private method unionOfDestinationEvents.

Thanks Alex!

braedenkloke committed:
docs: add todo item

Refactor docs of setFloorRequests

braedenkloke committed:
refactor(scheduler): remove setFloorRequests

Not needed now that nested test class is implemented.
Simply can set floorRequests by this.floorRequests = requests.

braedenkloke committed:
refactor(scheduler.isElevatorStopping): remove deprecated method filterDestinationEvents

Not needed. Simply call e.currentFloor() and pass to DestinationEvent constructor
to compare against union of destination sets.

braedenkloke committed:
feat: stub method getElevatorDirection

Encapsulates logic for getting elevator direction from a elevator state event.
High potential for reuse.

braedenkloke committed:
docs(scheduler): add @return to getElevatorDirection doc

braedenkloke committed:
Merge branch 'develop/latest' into develop/scheduler-check-is-elevator-stopping

braedenkloke committed:
feat(Scheduler.getElevatorDirection): implement method without error checking

braedenkloke committed:
docs(scheduler): remove outdated comment

braedenkloke committed:
docs(scheduler): add todo items for getElevatorDirection

braedenkloke committed:
docs: clarified various java docs

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -1,39 +1,81 @@
+    private Direction getElevatorDirection(ElevatorStateEvent e) {
+        // TODO(@braeden): should always return a valid direction or throw error
+        //   I'm worried about returning a null value!
+        // TODO(@braeden): should also assert that every direction is same?
```
Honestly, I thought that's how it was already tbh. idk if you want to implement that @ArthurAtangana?

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -1,39 +1,81 @@
+     * @return Direction elevator is travelling.
+     */
+    private Direction getElevatorDirection(ElevatorStateEvent e) {
+        // TODO(@braeden): should always return a valid direction or throw error
```
Nulls are valid use cases... but they have to be processed differently by the callers.

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -43,7 +85,7 @@ private void processElevatorEvent(ElevatorStateEvent event) {
             new Thread(new Loader(event, transmitterToFloor, transmitterToElevator)).start();
         else // Keep moving
-            transmitterToElevator.send(new MoveElevatorCommand(event.elevatorNum()));
+            transmitterToElevator.send(new MoveElevatorCommand(event.elevatorNum(), getElevatorDirection(event)));
```
Null case: Elevator has no passengers, serves highest priority floor request.
Highest priority floor request -> could look for oldest floor request? (This requires quite a few changes, leave as is for this PR but it's a blocking feature for this iteration)

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -1,39 +1,81 @@
+        union.addAll(floorRequests);
+        union.addAll(e.passengerCountMap().keySet());
+
+        return union.contains(new DestinationEvent(e.currentFloor(), getElevatorDirection(e)));
```
Null direction use case: No more passengers, only care to check highest priority direction (see comment on line 88).

orselane commented on src/SchedulerSubsystem/Scheduler.java:
```diff
@@ -72,4 +114,51 @@ public void run() {
+     *
+     * See README section Test for additional context.
+     */
+    public static class SchedulerTest {
```
I don't have time to research if this is good or not, I'm not reviewing this part.
